R/pareto_smooth.R

```diff
@@ -0,0 +1,326 @@
+#' Pareto khat diagnostic
+#'
+#' Estimate Pareto k value from Pareto smoothing the tail(s) of x. For
+#' further details see Vehtari et al. (2022).
+#'
+#' @template args-pareto
+#' @template args-methods-dots
+#' @return numeric vector of Pareto smoothing diagnostics
+#'
+#' @examples
+#' mu <- extract_variable_matrix(example_draws(), "mu")
+#' pareto_khat(mu)
+#'
+#' d <- as_draws_rvars(example_draws("multi_normal"))
+#' pareto_khat(d$Sigma)
+#' @export
+pareto_khat <- function(x, ...) UseMethod("pareto_khat")
+
+#' @rdname pareto_khat
+#' @export
+pareto_khat.default <- function(x,
+                                ...) {
+  pareto_smooth.default(x, return_smoothed = FALSE, ...)
+}
+
+#' @rdname pareto_khat
+#' @export
+pareto_khat.rvar <- function(x, ...) {
+  unlist(
+    summarise_rvar_by_element_with_chains(x, pareto_khat.default, ...),
+    recursive = FALSE
+  )
+}
+
+#' Pareto smoothing
+#'
+#' Smooth the tail(s) of x by fitting a generalized Pareto
+#' distribution. For further details see Vehtari et al. (2022).
+#'
+#' @template args-pareto
+#' @template args-methods-dots
+#' @param return_smoothed (logical) Should the smoothed x be returned?
+#'   If `TRUE`, returns x with smoothed tail(s). If `FALSE`, acts the
+#'   same as `pareto_khat`. Default is `TRUE`.
+#' @return List containing vector of smoothed values and vector of
+#'   pareto smoothing diagnostics.
+#'
+#' @examples
+#' mu <- extract_variable_matrix(example_draws(), "mu")
+#' pareto_smooth(mu)
+#'
+#' d <- as_draws_rvars(example_draws("multi_normal"))
+#' pareto_smooth(d$Sigma)
+#' @export
+pareto_smooth <- function(x, ...) UseMethod("pareto_smooth")
+
+#' @rdname pareto_smooth
+#' @export
+pareto_smooth.rvar <- function(x, ...) {
+    unlist(summarise_rvar_by_element_with_chains(x, pareto_smooth.default, ...), recursive = FALSE)
+}
+
+#' @rdname pareto_smooth
+#' @export
+pareto_smooth.default <- function(x,
```

Comment:
Is this function for a single variable or for a set of variables (a draws object) where we loop over the variables to return a vector of outputs? I recommend we have both. The default should be for a single variable. 

(Currently rhat and friends only support a single variable, but I think we should also write some versions for draws objects returning vectors there.)

Reply:
This default function is for a single variable, just like `rhat`, `ess_*` and others (for use in `summarise_draws`). I agree that it would be good to add methods for `draws` objects to  address #278, but maybe this should be done in another PR for all diagnostics?

Reply:
Another PR for the multi variable case makes sense indeed.